### ISSUE TITLE
FInd return cursors in embedded docs

### ISSUE THREAD

**Issue by maheshsetti** [Automattic/mongoose#6572]
**Do you want to request a *feature* or report a *bug*?**
I'm not sure whats wrong with the current behavior but its not feature.

**What is the current behavior?**
I have a schema with nested embedded docs, after find excu im getting cursor along with embedded docs like below

> { parameters:
   [ { parameterField: 'babyAge',
       parameterName: 'Baby Age',
       parameterUnit: 'Day',
       parameterCondition: 'low',
       parameterValue: '15',
       _id: 5b1665d90f7c6c19959b9b35 } ],
  cutoffCondition: 'equal',
  unit: 'ng/mL',
  cutoffVal: '100',
  _id: 5b1665d90f7c6c19959b9b34 }
{ parameters:
   [ { parameterField: 'babyAge',
       parameterName: 'Baby Age',
       parameterUnit: 'Day',
       parameterCondition: 'high',
       parameterValue: '15',
       _id: 5b1665d90f7c6c19959b9b33 } ],
  cutoffCondition: 'equal',
  unit: 'ng/mL',
  cutoffVal: '50',
  _id: 5b1665d90f7c6c19959b9b32 }
analyticCutoffs
[Function: toBSON]
{}
{ _id: 5b1665d90f7c6c19959b9b31,
  name: '17 HYDROXYPROGESTERONE',
  abbreviation: '17OHP',
  analyteType: 'Assays',
  analyticCutoffs:
   [ { parameters: [Array],
       cutoffCondition: 'equal',
       unit: 'ng/mL',
       cutoffVal: '100',
       _id: 5b1665d90f7c6c19959b9b34 },
     { parameters: [Array],
       cutoffCondition: 'equal',
       unit: 'ng/mL',
       cutoffVal: '50',
       _id: 5b1665d90f7c6c19959b9b32 } ] }
[Function: _cast]
[Function: _markModified]
[Function: _registerAtomic]
[Function: $__getAtomics]
[Function: hasAtomics]
[Function: _mapCast]
[Function: push]
[Function: nonAtomicPush]
[Function: $pop]
[Function: pop]
[Function: $shift]
[Function: shift]
[Function: pull]
[Function: splice]
[Function: unshift]
[Function: sort]
[Function: addToSet]
[Function: set]
[Function: toObject]
[Function: inspect]
[Function: indexOf]
[Function: pull]
[Function: id]
[Function: create]
[Function: notify]
true
[]
DocumentArray {
  casterConstructor:
   { [Function: EmbeddedDocument]
     schema:
      Schema {
        obj: [Object],
        paths: [Object],
        aliases: {},
        subpaths: {},
        virtuals: [Object],
        singleNestedPaths: {},
        nested: {},
        inherits: {},
        callQueue: [],
        _indexes: [],
        methods: {},
        methodOptions: {},
        statics: {},
        tree: [Object],
        query: {},
        childSchemas: [Array],
        plugins: [Array],
        s: [Object],
        _userProvidedOptions: [Object],
        options: [Object],
        '$implicitlyCreated': true,
        '$globalPluginsApplied': true,
        _requiredpaths: [Array] },
     '$isArraySubdocument': true,
     domain: undefined,
     _events: undefined,
     _maxListeners: undefined,
     setMaxListeners: [Function: setMaxListeners],
     getMaxListeners: [Function: getMaxListeners],
     emit: [Function: emit],
     addListener: [Function: addListener],
     on: [Function: addListener],
     prependListener: [Function: prependListener],
     once: [Function: once],
     prependOnceListener: [Function: prependOnceListener],
     removeListener: [Function: removeListener],
     removeAllListeners: [Function: removeAllListeners],
     listeners: [Function: listeners],
     listenerCount: [Function: listenerCount],
     eventNames: [Function: eventNames],
     options: { type: [Array] },
     path: 'analyticCutoffs',
     '$appliedMethods': true,
     '$appliedHooks': true },
  caster:
   { [Function: EmbeddedDocument]
     schema:
      Schema {
        obj: [Object],
        paths: [Object],
        aliases: {},
        subpaths: {},
        virtuals: [Object],
        singleNestedPaths: {},
        nested: {},
        inherits: {},
        callQueue: [],
        _indexes: [],
        methods: {},
        methodOptions: {},
        statics: {},
        tree: [Object],
        query: {},
        childSchemas: [Array],
        plugins: [Array],
        s: [Object],
        _userProvidedOptions: [Object],
        options: [Object],
        '$implicitlyCreated': true,
        '$globalPluginsApplied': true,
        _requiredpaths: [Array] },
     '$isArraySubdocument': true,
     domain: undefined,
     _events: undefined,
     _maxListeners: undefined,
     setMaxListeners: [Function: setMaxListeners],
     getMaxListeners: [Function: getMaxListeners],
     emit: [Function: emit],
     addListener: [Function: addListener],
     on: [Function: addListener],
     prependListener: [Function: prependListener],
     once: [Function: once],
     prependOnceListener: [Function: prependOnceListener],
     removeListener: [Function: removeListener],
     removeAllListeners: [Function: removeAllListeners],
     listeners: [Function: listeners],
     listenerCount: [Function: listenerCount],
     eventNames: [Function: eventNames],
     options: { type: [Array] },
     path: 'analyticCutoffs',
     '$appliedMethods': true,
     '$appliedHooks': true },
  '$isMongooseArray': true,
  path: 'analyticCutoffs',
  instance: 'Array',
  validators: [],
  setters: [],
  getters: [],
  options: { type: [ [Object] ] },
  _index: null,
  defaultValue: [Function],
  schema:
   Schema {
     obj:
      { parameters: [Array],
        cutoffCondition: [Object],
        cutoffVal: [Object],
        upperCutoffVal: [Object],
        lowerCutoffVal: [Object],
        unit: [Object] },
     paths:
      { parameters: [Object],
        cutoffCondition: [Object],
        cutoffVal: [Object],
        upperCutoffVal: [Object],
        lowerCutoffVal: [Object],
        unit: [Object],
        _id: [Object] },
     aliases: {},
     subpaths: {},
     virtuals: { id: [Object] },
     singleNestedPaths: {},
     nested: {},
     inherits: {},
     callQueue: [],
     _indexes: [],
     methods: {},
     methodOptions: {},
     statics: {},
     tree:
      { parameters: [Array],
        cutoffCondition: [Object],
        cutoffVal: [Object],
        upperCutoffVal: [Object],
        lowerCutoffVal: [Object],
        unit: [Object],
        _id: [Object],
        id: [Object] },
     query: {},
     childSchemas: [ [Object] ],
     plugins: [ [Object], [Object], [Object], [Object], [Object] ],
     s: { hooks: [Object] },
     _userProvidedOptions: { minimize: true, typeKey: 'type', strict: true },
     options:
      { minimize: true,
        typeKey: 'type',
        strict: true,
        id: true,
        noVirtualId: false,
        _id: true,
        noId: false,
        validateBeforeSave: true,
        read: null,
        shardKey: null,
        autoIndex: null,
        discriminatorKey: '__t',
        versionKey: '__v',
        capped: false,
        bufferCommands: true },
     '$implicitlyCreated': true,
     '$globalPluginsApplied': true,
     _requiredpaths: [ 'unit', 'cutoffCondition' ] },
  schemaOptions: {},
  '$isMongooseDocumentArray': true }
{ isNew: [Function: notify], save: [Function: notify] }

**If the current behavior is a bug, please provide the steps to reproduce.**
I have a async getlist method
![image](https://user-images.githubusercontent.com/9315052/41185075-c17de64e-6b95-11e8-9152-7ac05c5bf89b.png)

Then im calling async method and getting list of docs 
![image](https://user-images.githubusercontent.com/9315052/41185093-e4421358-6b95-11e8-8895-32f6c52836a3.png)

**What is the expected behavior?**
When I do `console.log(dbAnalytes[0].analyticCutoffs[0])`
It should console
![image](https://user-images.githubusercontent.com/9315052/41185130-4bc5bb42-6b96-11e8-85af-435dca46cc9a.png)

I'm not sure why its printing all mongoose cursor functions along with embedded docs
**Please mention your node.js, mongoose and MongoDB version.**
Node.js - v8.11.2
Mongoose.js - v5.1.4
MongoDB - v3.2.20

**Comment by lineus** [comment]
@maheshsetti it's very difficult to help without a repro script that is minimally complete. Whatever is going on in your code isn't readily apparent from the images, although the `getListAsyc` function doesn't need the async/await, as you can just return the promise.

Here is an example of a minimally complete repro script, please modify it or create one like it to demonstrate the issue you are reporting:

### 6572.js
```js
#!/usr/bin/env node
'use strict';

const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const schema = new Schema({
  name: String
});

const Test = mongoose.model('test', schema);

const test = new Test({
  name: 'Bird'
});

function go() {
  return Test.find();
}

async function run() {
  await conn.dropDatabase();
  await test.save();
  let found = await go();
  console.log(found);
  return conn.close();
}

run();
```
### Output: 
```
issues: ./6572.js
[ { _id: 5b1bb9a14ba81b5d61caf0ab, name: 'Bird', __v: 0 } ]
issues:
```

**Comment by maheshsetti** [comment]
The following arrow marked arrays are `CoreMongooseArray`
![image](https://user-images.githubusercontent.com/9315052/41191385-cb695958-6bff-11e8-87f0-1011258a2fea.png)

I have schema like below
![image](https://user-images.githubusercontent.com/9315052/41191414-15e56cc4-6c00-11e8-9cc7-68457f1d2f3b.png)

Save and find are fine, I think the problem with `CoreMongooseArray` class. When i try to ready embedded docs `analyticCutoffs` it showing `CoreMongooseArray` as `parameters`

**Comment by lineus** [comment]
@maheshsetti thanks for adding your schema. I tried to reproduce what you are seeing 

> all mongoose cursor functions along with embedded docs

 with the following script:

### 6572.js
```
#!/usr/bin/env node
'use strict';

const mongoose = require('mongoose');
mongoose.connect('mongodb://localhost/test');
const conn = mongoose.connection;
const Schema = mongoose.Schema;

const albumSchema = new Schema({
  title: String,
  players: [{
    name: String,
    instrument: [{
      kind: String
    }]
  }]
});

const Album = mongoose.model('album', albumSchema);

const test = new Album({
  title: 'At The Blackhawk',
  players: [
    { name: 'Rouse', instrument: [{ kind: 'Sax' }, { kind: 'Tenor Sax' }]},
    { name: 'Monk', instrument: [{ kind: 'Piano' }] }
  ]
});

function go() {
  return Album.find();
}

async function run() {
  await conn.dropDatabase();
  await test.save();
  let found = await go();
  console.log(found[0].players[0]);
  return conn.close();
}

run();
```

### Output:
```
issues: ./6572.js
{ instrument:
   [ { _id: 5b1bdb64258dd66b47aea4c8, kind: 'Sax' },
     { _id: 5b1bdb64258dd66b47aea4c7, kind: 'Tenor Sax' } ],
  _id: 5b1bdb64258dd66b47aea4c6,
  name: 'Rouse' }
issues:
```

Can you please either create a reproducible example or modify mine such that it demonstrates the issue?

**Comment by maheshsetti** [comment]
@lineus I have reproduced by using following script

```javascript
let found = await go();
console.log(found[0].players[0]);
console.log(found[0].players[0].instrument);
let arInst = found[0].players[0].instrument;
for(let j in arInst){
	console.log(arInst[j]);
}
```
***Output***
![image](https://user-images.githubusercontent.com/9315052/41193482-86c603a0-6c1d-11e8-8fc7-a18a30e48199.png)

**Comment by lineus** [comment]
@maheshsetti  you can use `for..of` instead,

```js
  let arInst = found[0].players[0].instrument;
  for (let j of arInst) {
    console.log(j); // note the value of j is the value not the key.
  }
```
### Outputs:
```
issues: ./6572.js
{ _id: 5b1bf9803f55c578945e108f, kind: 'Sax' }
{ _id: 5b1bf9803f55c578945e108e, kind: 'Tenor Sax' }
issues:
```
@vkarpov15 do you think we should make the methods non-enumerable on MongooseArray/CoreMongooseArray ? Or is the answer here just use `for..of`?

**Comment by lineus** [comment]
@vkarpov15 I'm just wrapping up the new tests for a PR that makes all of the properties non-enumerable if that's the direction you want to go in.

**Comment by maheshsetti** [comment]
@lineus  
`for..of ` works fine. It would be scalable feature if `MongooseArray/CoreMongooseArray` supports `non-enumerable` type keys.

**Comment by vkarpov15** [comment]
Fixed will be in 5.6.6, although in general you [shouldn't use `for/in` to iterate through an array](https://thecodebarbarian.com/for-vs-for-each-vs-for-in-vs-for-of-in-javascript.html).
